More memory changes with some thoughts on error handling

gurchetansingh opened the pull request:
* [x] Added an entry to `CHANGELOG_VULKANO.md` or `CHANGELOG_VK_SYS.md` if knowledge of this change could be valuable to users
* [ ] Updated documentation to reflect any user-facing changes - in this repository
* [ ] Updated documentation to reflect any user-facing changes - PR to the [guide](https://github.com/vulkano-rs/vulkano-www) that fixes existing documentation invalidated by this PR.
* [ x] Ran `cargo fmt` on the changes

Hey, I know the tests don't pass yet, but wondering for some feedback on the VulkanoResult.

CC: @hakolao @Rua @Eliah-Lakhin @maximbaz

Rua commented:
I don't think using `VulkanoResult` like this is a very good idea. Having an explicit error type means the user knows in advance what errors a function is going to return, and they can exhaustively match on it. Now every function can return every error, so it's much less clear and the user has to cover every possible variant. What was the motivation for making this change?

Aside from that, in the Ash backend conversion I've been working on, the `check_errors` function and the `Error` and `Success` enums disappear. `ash::prelude::VkResult` converts directly to each individual error type a function returns. For example:
```Rust
impl From<ash::vk::Result> for InstanceCreationError {
    #[inline]
    fn from(err: ash::vk::Result) -> Self {
        use ash::vk::Result as R;
        match err {
            R::ERROR_OUT_OF_HOST_MEMORY | R::ERROR_OUT_OF_DEVICE_MEMORY => {
                OomError::from(err).into()
            }
            R::ERROR_INITIALIZATION_FAILED => Self::InitializationFailed,
            R::ERROR_LAYER_NOT_PRESENT => Self::LayerNotPresent,
            R::ERROR_EXTENSION_NOT_PRESENT => Self::ExtensionNotPresent,
            R::ERROR_INCOMPATIBLE_DRIVER => Self::IncompatibleDriver,
            _ => panic!("unexpected error: {:?}", err),
        }
    }
}```
This method could be used for Vulkano's current state as well, converting from `vk::Result`. You simply place a `?` after the called function. You just need to pay attention when the function could return a value that is neither `SUCCESS` nor an `ERROR`, e.g. `vkAcquireNextImageKHR`.

gurchetansingh commented:
Thank you rua for your feedback.  I have dropped the Vulkano result from this series and look forward to your Ash change, as I view it fundamental to the project going forward!!  

I guess my thinking is I just don't see the overall structure for Vulkano error handling yet.  Add details in the commit message and the commit.  I think if we want to be a validation layer we should just return the VUID violation and in theory the user can match on that.  I think it can be simplified and improved, but of course I'm just raising the issue and would welcome other solutions.  For now, it's more a "quality of life" thing and not a priority, so dropped.

Eliah-Lakhin commented:
@gurchetansingh Thank you for your contribution. Please let me know when it's ready for final review.

Regarding VulkanoResult/Error topic. I would keep error system API as it is, without changes.

gurchetansingh commented:
@Eliah-Lakhin ready for final review now.

gurchetansingh commented:
One thing to note is we no longer use `memory_type: MemoryType` in the builder.  The reason is I think we should autogenerate most of these builders as well [after the Ash change], so working with raw types and then validating later was chosen.

Eliah-Lakhin commented:
@gurchetansingh Sorry for delaying. I usually don't have enough time to manage Pull Requests properly during the week, and I left this job for my days off. So I will take a closer look on Sat-Sun.

For now, can you explain:

1. What is the purpose of this change?
2. Why do we need to remove `MemoryType` and how is it connected to the PR's context? If this are irrelevant topics to the main PR's task context, maybe consider moving to a different Pull Request?
3. If there are breaking changes, these changes need to be highlighted in the changelogs.

gurchetansingh commented:
@Eliah-Lakhin thank you for taking time to contribute this project!

1)  For my use case, I need to:

* Import external memory from one process to another.
* A wrapper around `DeviceMemoryMapping`, so I can pass the raw pointer to the hypervisor.
* Also a cross-beam simplification, which makes Vulkano integration easier.
 
Take a look at [crrev.com/c/2786286](crrev.com/c/2786286) for the use case [I've modified it a bit, but the Vulkano side remains unchanged]

2) It's just a minor inconsequential detail -- anyone using `DeviceMemoryBuilder` already can just use memory_type.id().  The bigger issue is we're mixing Vulkano objects (`MemoryType`) and the builders -- which will prevent auto-generated builders around Ash in the future.  But you are right, that's too far ahead and opaque to a reviewer such as yourself 🙂.

Since getting things merged can take a while, I do tend to put multiple discrete changes in one MR.  In future, I definitely agree smaller, easier to review MRs are better ... I hope the community gets there!

3) Done, added a changelog entry for the minor `DeviceMemoryBuilder` breaking change!

Eliah-Lakhin commented on CHANGELOG_VULKANO.md:
```diff
@@ -1,4 +1,7 @@
 # Unreleased
+- **Breaking ** DeviceMemoryBuilder::new() takes in `memory_index` rather than `MemoryType`.
```
Extra whitespace

Eliah-Lakhin commented on vulkano/src/memory/device_memory.rs:
```diff
@@ -87,28 +112,21 @@ impl<'a> DeviceMemoryBuilder<'a> {
-            memory_type.physical_device().internal_object()
-        );
-
+    pub fn new(device: Arc<Device>, memory_index: u32, size: usize) -> DeviceMemoryBuilder<'a> {
```
Is this method still safe? What would happen if the user passes incompatible value of `memory_index`?

gurchetansingh replied:
Yes, it is.  Validation is now done in `DeviceMemoryBuilder::build()`, since on the fly validation while adding to the builder I feel does not scale or lend itself to autogeneration (say we a ImageBuilder or BufferBuilder in the future).

Eliah-Lakhin commented on vulkano/src/memory/device_memory.rs:
```diff
@@ -87,28 +112,21 @@ impl<'a> DeviceMemoryBuilder<'a> {
     ///
     /// - Panics if `size` is 0.
     /// - Panics if `memory_type` doesn't belong to the same physical device as `device`.
```
`memory_type` argument was replaced, so this comment needs to be corrected accordingly.

gurchetansingh replied:
Fixed.

Eliah-Lakhin commented on vulkano/src/memory/device_memory.rs:
```diff
@@ -120,38 +138,28 @@ impl<'a> DeviceMemoryBuilder<'a> {
-            } else {
-                self.allocate.pNext = ptr;
-            }
+        if self.dedicated_info.is_some() {
```
Why assertion replaced with explicit panic?

gurchetansingh replied:
Fixed.

Eliah-Lakhin commented on vulkano/src/memory/device_memory.rs:
```diff
@@ -160,67 +168,155 @@ impl<'a> DeviceMemoryBuilder<'a> {
-        let handle_bits = handle_types.to_bits();
-        if handle_bits & vk::EXTERNAL_MEMORY_HANDLE_TYPE_DMA_BUF_BIT_EXT != 0 {
-            assert!(self.device.loaded_extensions().ext_external_memory_dmabuf);
+        if self.export_info.is_some() {
```
Why assertion replaced with explicit panic?

gurchetansingh replied:
Fixed.

Eliah-Lakhin reviewed (changes requested):
@gurchetansingh Thank you for detailed explanation. I would like to proceed with merging of this PR, but I left a few questions/comments to address first. Mostly cosmetic.

gurchetansingh commented:
@Eliah-Lakhin thank you for the review!

Eliah-Lakhin commented:
@gurchetansingh Pull Request merged. Thank you for your contribution and for the fixes!
